fem/bilininteg_hcurl.cpp

```diff
@@ -1802,6 +1802,536 @@ void PAHcurlH1Apply3D(const int D1D,
+            {
+               const int local_index = c_dofs1D * o_dofs1D + ey*c_dofs1D + ex;
+               y(local_index, e) += B(ex, dx) * vw(dx, ey);
+            }
+         }
+      }
+   });
+
+}
+
+static void PAHcurlApplyGradientTranspose2D(
+   const int c_dofs1D, const int o_dofs1D, const int NE,
+   const Array<double> &_B, const Array<double> &_G,
+   const Vector &_x, Vector &_y)
+{
+   auto B = Reshape(_B.Read(), c_dofs1D, c_dofs1D);
+   auto G = Reshape(_G.Read(), o_dofs1D, c_dofs1D);
+
+   auto x = Reshape(_x.Read(), 2 * c_dofs1D * o_dofs1D, NE);
+   auto y = Reshape(_y.ReadWrite(), c_dofs1D, c_dofs1D, NE);
+
+   Vector hwork(c_dofs1D * o_dofs1D);
+   auto hw = Reshape(hwork.ReadWrite(), c_dofs1D, o_dofs1D);
+
+   Vector vwork(c_dofs1D * c_dofs1D);
+   auto vw = Reshape(vwork.ReadWrite(), c_dofs1D, c_dofs1D);
+
+   MFEM_FORALL(e, NE,
+   {
+      // horizontal part (open x, closed y)
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int ex = 0; ex < o_dofs1D; ++ex)
+         {
+            hw(dy, ex) = 0.0;
+            for (int ey = 0; ey < c_dofs1D; ++ey)
+            {
+               const int local_index = ey*o_dofs1D + ex;
+               hw(dy, ex) += B(ey, dy) * x(local_index, e);
+            }
+         }
+      }
+
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int dx = 0; dx < c_dofs1D; ++dx)
+         {
+            for (int ex = 0; ex < o_dofs1D; ++ex)
+            {
+               y(dx, dy, e) += G(ex, dx) * hw(dy, ex);
+            }
+         }
+      }
+
+      // vertical part (open y, closed x)
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int ex = 0; ex < c_dofs1D; ++ex)
+         {
+            vw(dy, ex) = 0.0;
+            for (int ey = 0; ey < o_dofs1D; ++ey)
+            {
+               const int local_index = c_dofs1D * o_dofs1D + ey*c_dofs1D + ex;
+               vw(dy, ex) += G(ey, dy) * x(local_index, e);
+            }
+         }
+      }
+
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int dx = 0; dx < c_dofs1D; ++dx)
+         {
+            for (int ex = 0; ex < c_dofs1D; ++ex)
+            {
+               y(dx, dy, e) += B(ex, dx) * vw(dy, ex);
+            }
+         }
+      }
+   });
+}
+
+static void PAHcurlApplyGradient3D(const int c_dofs1D,
+                                   const int o_dofs1D,
+                                   const int NE,
+                                   const Array<double> &_B,
+                                   const Array<double> &_G,
+                                   const Vector &_x,
+                                   Vector &_y)
+{
+   auto B = Reshape(_B.Read(), c_dofs1D, c_dofs1D);
+   auto G = Reshape(_G.Read(), o_dofs1D, c_dofs1D);
+
+   auto x = Reshape(_x.Read(), c_dofs1D, c_dofs1D, c_dofs1D, NE);
+   auto y = Reshape(_y.ReadWrite(), (3 * c_dofs1D * c_dofs1D * o_dofs1D), NE);
+
+   Vector pxwork1(c_dofs1D * c_dofs1D * c_dofs1D);
```

Comment:
In existing PA kernels in MFEM, usually arrays like this are defined with something like `double a[m][n][p]` inside the `MFEM_FORALL`. Is there a difference between that and the `ReadWrite` used here, in terms of how the memory is allocated and used?

Reply:
Probably? The Nedelec dofs come in three separate groups (x, y, z directions), and each one of those groups has size `c_dofs * c_dofs * o_dofs`. I wasn't sure the best way to organize that, I chose this and got it to work, but probably there's more efficient ways for memory access etc.

Reply:
Oh, I misunderstood your comment (thought you were talking about line 1966). You are right and I should probably change the way `pxwork1` is allocated.

Reply:
I am not familiar enough with the memory management to know the details of the difference, so it is unclear to me whether a change is needed. The GPU experts should have more insight (@camierjs @YohannDudouit ).

Reply:
I agree with @dylan-copeland you should declare these inside the `MFEM_FORALL`, currently this generates a lot a memory allocations/deallocations which are very expensive on GPU. I'm also suspicious that this code works on GPU, because all the threads compute in the same memory space. If you declare the the `Vector` inside the `MFEM_FORALL` using arrays, then the memory space is private to each thread.